apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabTheorycraft/ArtifactSubCard/ArtifactSubstatEditor.tsx

```diff
@@ -0,0 +1,204 @@
+  const substatValue = getSubstatValue(statKey, rarity, substatsType)
+  const [rolls, setRolls] = useState(() => value / substatValue)
+  useEffect(() => setRolls(value / substatValue), [value, substatValue])
+
+  const unit = KeyMap.unit(statKey)
+  const displayValue = rolls * substatValue
+
+  const rv = ((rolls * substatValue) / getSubstatValue(statKey)) * 100
+  const numMains = mainStatKeys.reduce(
+    (t, ms) => t + (ms === statKey ? 1 : 0),
+    0
+  )
+  const maxRolls = (5 - numMains) * (artMaxLevel[rarity] / 4 + 1)
+  // 0.0001 to nudge float comparasion
+  const invalid = rolls - 0.0001 > maxRolls
+  const setRValue = useCallback(
+    (r: number) => setValue(r * substatValue),
+    [setValue, substatValue]
+  )
+
+  return (
+    <Stack spacing={0.5}>
+      <Box
+        display="flex"
+        gap={1}
+        justifyContent="space-between"
+        alignItems="center"
+      >
+        <CustomNumberInput
+          color={displayValue ? 'success' : 'primary'}
+          float={KeyMap.unit(statKey) === '%'}
+          endAdornment={
+            KeyMap.unit(statKey) || <Box width="1em" component="span" />
+          }
+          value={parseFloat(displayValue.toFixed(2))}
+          onChange={(v) => v !== undefined && setValue(v)}
+          sx={{ borderRadius: 1, px: 1, height: '100%', width: '5em' }}
+          inputProps={{ sx: { textAlign: 'right' }, min: 0 }}
+          disabled={disabled}
+        />
+        <CardDark
+          sx={{
+            p: 0.5,
+            minWidth: '11em',
+            flexGrow: 1,
+            display: 'flex',
+            gap: 1,
+            alignItems: 'center',
+            justifyContent: 'center',
+          }}
+        >
+          <StatIcon statKey={statKey} iconProps={{ fontSize: 'inherit' }} />
+          {KeyMap.getStr(statKey)}
+          {KeyMap.unit(statKey)}
+        </CardDark>
+        <CustomNumberInput
+          color={value ? (invalid ? 'warning' : 'success') : 'primary'}
+          float
+          startAdornment={
+            <Box
+              sx={{
+                whiteSpace: 'nowrap',
+                width: '6em',
+                display: 'flex',
+                justifyContent: 'space-between',
+              }}
+            >
+              <span>
+                {artDisplayValue(substatValue, unit)}
+                {unit}
+              </span>
+              <span>x</span>
+            </Box>
+          }
+          value={parseFloat(rolls.toFixed(2))}
+          onChange={(v) => v !== undefined && setValue(v * substatValue)}
+          sx={{ borderRadius: 1, px: 1, my: 0, height: '100%', width: '6.5em' }}
+          inputProps={{ sx: { textAlign: 'right', pr: 0.5 }, min: 0, step: 1 }}
+          disabled={disabled}
+        />
+        {/* <BootstrapTooltip title={<Typography>{t(numMains ? `tabTheorycraft.maxRollsMain` : `tabTheorycraft.maxRolls`, { value: maxRolls })}</Typography>} placement="top"> */}
```

Comment:
Why is this commented out? It might have been like that before the refactor idk

Reply:
it was popping up in locations that conflicts with max input. move it to a smaller icon.